### PR TITLE
managed private networks: update the create and list commands

### PR DESCRIPTION
**Pull request by mcorbin**
#319 

- Add the `startip`, `endip`, `netmask` parameters to the create
  network command
- Show the `startip`, `endip`, `netmask` fields in the list networks command

### CODE REVIEW

File: networks.go [hunk 1]
```diff
@@ -114,18 +116,18 @@ type CreateNetwork struct {
...
 	EndIpv6           net.IP `json:"endipv6,omitempty" doc:"the ending IPv6 address in the IPv6 network range"`
 	Gateway           net.IP `json:"gateway,omitempty" doc:"the gateway of the network. Required for Shared networks and Isolated networks when it belongs to VPC"`
 	IP6CIDR           *CIDR  `json:"ip6cidr,omitempty" doc:"the CIDR of IPv6 network, must be at least /64"`
 	IP6Gateway        net.IP `json:"ip6gateway,omitempty" doc:"the gateway of the IPv6 network. Required for Shared networks and Isolated networks when it belongs to VPC"`
 	IsolatedPVlan     string `json:"isolatedpvlan,omitempty" doc:"the isolated private vlan for this network"`
 	Name              string `json:"name,omitempty" doc:"the name of the network"` // This field is required but might be empty
-	Netmask           net.IP `json:"netmask,omitempty" doc:"the netmask of the network. Required for Shared networks and Isolated networks when it belongs to VPC"`
+	Netmask           net.IP `json:"netmask,omitempty" doc:"the netmask of the network.  Required for managed networks"`
```
Comment (greut) [inline networks.go]:
```console
$ go run generate/main.go -apis listApis.json -cmd CreateNetwork

netmask: wrong doc want "the netmask of the network. Required for managed networks." got "the netmask of the network.  Required for managed networks"
```

File: networks.go [hunk 2]
```diff
@@ -30,7 +31,7 @@ type Network struct {
 	IsPersistent                bool          `json:"ispersistent,omitempty" doc:"list networks that are persistent"`
 	IsSystem                    bool          `json:"issystem,omitempty" doc:"true if network is system, false otherwise"`
 	Name                        string        `json:"name,omitempty" doc:"the name of the network"`
-	Netmask                     net.IP        `json:"netmask,omitempty" doc:"the network's netmask"`
+	Netmask                     net.IP        `json:"netmask,omitempty" doc:"the network's netmask. Required for managed networks"`
```
Comment (greut) [inline networks.go]: this doesn't match the content of the `listApis`

File: networks.go [hunk 3]
```diff
@@ -155,10 +157,13 @@ type UpdateNetwork struct {
 	CustomID          *UUID  `json:"customid,omitempty" doc:"an optional field, in case you want to set a custom id to the resource. Allowed to Root Admins only"`
 	DisplayNetwork    *bool  `json:"displaynetwork,omitempty" doc:"an optional field, whether to the display the network to the end user or not."`
 	DisplayText       string `json:"displaytext,omitempty" doc:"the new display text for the network"`
+	EndIP             net.IP `json:"endip,omitempty" doc:"the ending IP address in the network IP range. Required for managed networks."`
 	GuestVMCIDR       *CIDR  `json:"guestvmcidr,omitempty" doc:"CIDR for Guest VMs,Cloudstack allocates IPs to Guest VMs only from this CIDR"`
 	Name              string `json:"name,omitempty" doc:"the new name for the network"`
+	Netmask           net.IP `json:"netmask,omitempty" doc:"the netmask of the network.  Required for managed networks"`
```
Comment (greut) [inline networks.go]:
```console
$ go run generate/main.go -apis listApis.json -cmd UpdateNetwork

netmask: wrong doc want "the netmask of the network. Required for managed networks." got "the netmask of the network.  Required for managed networks"
```

File: cmd/exo/cmd/privnet_create.go [hunk 4]
```diff
@@ -83,12 +96,25 @@ func privnetCreate(name, desc, zoneName string) error {
...
+	var startip, endip, netmask net.IP
+	if startIPAddr != "" {
+		startip = net.ParseIP(startIPAddr)
+	}
+	if endIPAddr != "" {
+		endip = net.ParseIP(endIPAddr)
+	}
+	if netmaskAddr != "" {
+		netmask = net.ParseIP(netmaskAddr)
+	}
```
Comment (greut) [inline cmd/exo/cmd/privnet_create.go]: the `if` checks are not required.

File: cmd/exo/cmd/privnet_create.go [hunk 5]
```diff
@@ -102,15 +128,18 @@ func privnetCreate(name, desc, zoneName string) error {
 	newNet := resp.(*egoscale.Network)
 
 	table := table.NewTable(os.Stdout)
-	table.SetHeader([]string{"Name", "Description", "ID"})
-	table.Append([]string{newNet.Name, newNet.DisplayText, newNet.ID.String()})
+	table.SetHeader([]string{"Name", "Description", "ID", "StartIP", "EndIP", "Netmask"})
+	table.Append([]string{newNet.Name, newNet.DisplayText, newNet.ID.String(), newNet.StartIP.String(), newNet.EndIP.String(), newNet.Netmask.String()})
```
Comment (greut) [inline cmd/exo/cmd/privnet_create.go]: `<nil>` is an implementation details that shouldn't be shown.

| NAME | DESCRIPTION |                  ID                  | STARTIP | ENDIP | NETMASK |
|------|-------------|--------------------------------------|---------|-------|---------|
| bar  |             | d8283a3a-a9ff-4f5d-b385-f7ad16d4c311 | \<nil>   | \<nil> | \<nil>   |

File: cmd/exo/cmd/privnet_create.go [hunk 6]
```diff
@@ -83,12 +96,19 @@ func privnetCreate(name, desc, zoneName string) error {
 		return err
 	}
 
+	startip := net.ParseIP(startIPAddr)
+	endip := net.ParseIP(endIPAddr)
+	netmask := net.ParseIP(netmaskAddr)
```
Comment (greut) [inline cmd/exo/cmd/privnet_create.go]: :+1:

**Comment by mcorbin** [comment]
I rebased by branch.
I removed everything related to updateNetworks (and will move it in my other PR).
`go run generate/main.go -apis listApis.json -cmd CreateNetwork` is now OK.

The tables for createNetwork/listNetworks show empty values instead of <nil>.

File: cmd/exo/cmd/privnet_list.go [hunk 7]
```diff
@@ -59,7 +61,14 @@ func listPrivnets(zone string, table *table.Table) error {
...
 
-			table.Append([]string{zone, pn.Name, pn.ID.String(), vmNum})
+			table.Append([]string{
+				zone,
+				pn.Name,
+				pn.ID.String(),
+				vmNum,
+				utils.EmptyStringIfNilIP(pn.StartIP),
+				utils.EmptyStringIfNilIP(pn.EndIP),
+				utils.EmptyStringIfNilIP(pn.Netmask)})
```
[the remark this answers is not in the document]
Comment (mcorbin) [reply cmd/exo/cmd/privnet_list.go]: Done ;)

File: cmd/exo/cmd/privnet.go [hunk 8]
```diff
@@ -34,3 +37,13 @@ func getNetworkByName(name string) (*egoscale.Network, error) {
 func init() {
 	RootCmd.AddCommand(privnetCmd)
 }
+
+// getDhcpString returns the string representation for a DHCP
+func getDhcpString(startIP, endIP, netmask net.IP) string {
```
Comment (greut) [inline cmd/exo/cmd/privnet.go]: `s/getDhcpString/dhcpRange/`

https://golang.org/doc/effective_go.html#Getters

Comment (mcorbin) [reply cmd/exo/cmd/privnet.go]: done

**Comment by mcorbin** [comment]
I updated the privnet create command:

```
exo privnet create foobar --startip 10.0.0.20 --endip 10.0.0.30 --mask 24
┼────────┼─────────────┼──────────────────────────────────────┼─────────────────────────┼
│  NAME  │ DESCRIPTION │                  ID                  │          DHCP           │
┼────────┼─────────────┼──────────────────────────────────────┼─────────────────────────┼
│ foobar │             │ 3180f776-dcb6-42b7-9074-db9f7668a6ee │ 10.0.0.20-10.0.0.30 /24 │
┼────────┼─────────────┼──────────────────────────────────────┼─────────────────────────┼
```

**Comment by greut** [comment]
@mcorbin then it matches the doc, thanks.

```go
   Use: "create <name>",
```

File: cmd/exo/cmd/privnet_create.go [hunk 9]
```diff
@@ -102,15 +116,23 @@ func privnetCreate(name, desc, zoneName string) error {
...
 	table.Render()
 	return nil
 }
 
 func init() {
-	privnetCreateCmd.Flags().StringP("name", "n", "", "Private network name")
 	privnetCreateCmd.Flags().StringP("description", "d", "", "Private network description")
+	privnetCreateCmd.Flags().StringP("startip", "s", "", "the beginning IP address in the network IP range. Required for managed networks.")
+	privnetCreateCmd.Flags().StringP("endip", "e", "", "the ending IP address in the network IP range. Required for managed networks.")
+	privnetCreateCmd.Flags().StringP("mask", "m", "", "the netmask of the network.  Required for managed networks")
```
Comment (vincentbernat) [inline cmd/exo/cmd/privnet_create.go]: "mask", could be `UintP`?

Comment (mcorbin) [reply cmd/exo/cmd/privnet_create.go]: yes, updated ;) thanks

Comment (greut) [reply cmd/exo/cmd/privnet_create.go]: I'd still accept both though.

Comment (mcorbin) [reply cmd/exo/cmd/privnet_create.go]: You want `mask` and `netmask` ?

Comment (mcorbin) [reply cmd/exo/cmd/privnet_create.go]: Now, the command supports both.

Comment (greut) [reply cmd/exo/cmd/privnet_create.go]: @mcorbin the minimum is to be as good as the API, then we can do more but I wouldn't do things differently.

File: cmd/exo/cmd/privnet.go [hunk 10]
```diff
@@ -34,3 +37,23 @@ func getNetworkByName(name string) (*egoscale.Network, error) {
...
+func dhcpRange(startIP, endIP, netmask net.IP) string {
+	if startIP != nil && endIP != nil && netmask != nil {
+		mask := net.IPMask(netmask.To4())
+		prefixSize, _ := mask.Size()
+		return fmt.Sprintf("%s-%s /%d", startIP.String(), endIP.String(), prefixSize)
+	}
+	return "n/a"
+}
+
+func netmaskString(ip net.IP, mask uint) net.IP {
```
Comment (greut) [inline cmd/exo/cmd/privnet.go]: this is confusing: `func netmaskString -> net.IP`, return a string.

File: cmd/exo/cmd/privnet_create.go [hunk 11]
```diff
@@ -16,11 +17,27 @@ var privnetCreateCmd = &cobra.Command{
...
+		}
+		endip, err := cmd.Flags().GetString("endip")
+		if err != nil {
+			return err
+		}
+		mask, err := cmd.Flags().GetUint("mask")
+		if err != nil {
+			return err
+		}
+		netmask, err := cmd.Flags().GetString("netmask")
```
Comment (greut) [inline cmd/exo/cmd/privnet_create.go]: I'd keep only one of them, it's easier to support `255.255.255.0`, `/24` or `24` syntaxes.

File: networks.go [hunk 12]
```diff
@@ -114,18 +116,18 @@ type CreateNetwork struct {
...
 	EndIpv6           net.IP `json:"endipv6,omitempty" doc:"the ending IPv6 address in the IPv6 network range"`
 	Gateway           net.IP `json:"gateway,omitempty" doc:"the gateway of the network. Required for Shared networks and Isolated networks when it belongs to VPC"`
 	IP6CIDR           *CIDR  `json:"ip6cidr,omitempty" doc:"the CIDR of IPv6 network, must be at least /64"`
 	IP6Gateway        net.IP `json:"ip6gateway,omitempty" doc:"the gateway of the IPv6 network. Required for Shared networks and Isolated networks when it belongs to VPC"`
 	IsolatedPVlan     string `json:"isolatedpvlan,omitempty" doc:"the isolated private vlan for this network"`
 	Name              string `json:"name,omitempty" doc:"the name of the network"` // This field is required but might be empty
-	Netmask           net.IP `json:"netmask,omitempty" doc:"the netmask of the network. Required for Shared networks and Isolated networks when it belongs to VPC"`
+	Netmask           net.IP `json:"netmask,omitempty" doc:"the netmask of the network. Required for managed networks."`
```
Comment (greut) [inline networks.go]: could it be `net.IPMask`?

File: cmd/exo/cmd/privnet.go (hunk 10 above)
Comment (mcorbin) [reply cmd/exo/cmd/privnet.go]: Updated

File: networks.go (hunk 12 above)
Comment (mcorbin) [reply networks.go]: Nope, the API needs an IP.

Comment (greut) [reply networks.go]: we would have to create how own type, right. Not worth it.

File: cmd/exo/cmd/privnet_create.go (hunk 11 above)
Comment (mcorbin) [reply cmd/exo/cmd/privnet_create.go]: Ok, now the command supports only netmask (255.255.255.0)

File: cmd/exo/cmd/privnet_create.go [hunk 13]
```diff
@@ -35,12 +48,6 @@ var privnetCreateCmd = &cobra.Command{
 
 		if name == "" && zone == "" {
 			reader := bufio.NewReader(os.Stdin)
-			if name == "" {
-				name, err = readInput(reader, "Name", "")
-				if err != nil {
-					return err
-				}
-			}
```
Comment (greut) [inline cmd/exo/cmd/privnet_create.go]: `if zone == "" {` should be enough

File: cmd/exo/cmd/privnet_create.go [hunk 14]
```diff
@@ -16,11 +17,23 @@ var privnetCreateCmd = &cobra.Command{
 	Short:   "Create private network",
 	Aliases: gCreateAlias,
 	RunE: func(cmd *cobra.Command, args []string) error {
-		name, err := cmd.Flags().GetString("name")
+		if len(args) < 1 {
```
Comment (greut) [inline cmd/exo/cmd/privnet_create.go]: `if len(args) != 1` seems more accurate.

Comment (mcorbin) [reply cmd/exo/cmd/privnet_create.go]: Updated

File: cmd/exo/cmd/privnet_create.go (hunk 13 above)
Comment (mcorbin) [reply cmd/exo/cmd/privnet_create.go]: Updated

File: cmd/exo/cmd/privnet_create.go [hunk 15]
```diff
@@ -83,12 +90,19 @@ func privnetCreate(name, desc, zoneName string) error {
...
+
 	s := resp.(*egoscale.ListNetworkOfferingsResponse)
 
 	req := &egoscale.CreateNetwork{
 		DisplayText: desc,
 		Name:        name,
 		ZoneID:      zone,
+		StartIP:     startIP,
+		EndIP:       endIP,
+		Netmask:     netmaskIP,
```
Comment (greut) [inline cmd/exo/cmd/privnet_create.go]: You're relying on the API for doing the validation here, which might be okay.

**Pull request merged**